pup-0003.md

```diff
@@ -0,0 +1,154 @@
+---
+title: Cherry-Picking Commits
+author: David Davis
+created: 01-Apr-2017
+status:
+---
+
+## Summary
+
+Currently in Pulp, we merge features to master. However, hotfixes and bug fixes
+are merged to release branches which then are merged to newer release branches
+and then to master (a process called "merging forward"). More info about
+merging forward can be found in our documentation:
+
+http://docs.pulpproject.org/en/2.12/dev-guide/contributing/merging.html#merging-your-changes
+
+Also, a good illustration of this process is outlined in the following webpage:
+
+http://nvie.com/posts/a-successful-git-branching-model/
+
+This PUP is to use cherry-picking instead of merging forward. Using this model,
+PRs would be made against master and then changes would be cherry-picked to
+release branches as needed. For a more detailed workflow, see the Design
+section.
+
+
+## Motivation
+
+There are a few motivating factors for this proposal. First, our git history is
+complex with multiple branches feeding into one another and numerous merge
+commits. Using cherry-picks would simplify our git history by keeping it more
+linear and eliminating merge commits.
+
+Secondly, opening a PR requires knowledge of our current process. This can
+create a barrier to entry for community members. With the cherry-pick model,
+all PRs are opened against master and the onus to update release branches is on
+the person merging (who usually has better knowledge of which branches a commit
+needs to be applied to).
+
+Another minor caveat is that changes that we want to only be made to older
+release branches (but not to newer ones) still need to be merged forward. We
+could elminate the need for this with cherry-picking. It's worth noting that
+this rarely happens.
+
+Lastly, we've also seen some mistakes where branches are accidentally merged
+into other branches. Commits could be accidentally cherry-picked but this would
```

Comment:
Some extra sentences prior to the sentence starting with 'Commits'.

These situations have occurred multiple times, and each time required multiple people several hours to resolve. After being resolved the git history is convoluted with huge negative commits which later caused confusion at release time.